src/posts/2020-07-27-unsucking-your-project-onboarding.md

```diff
@@ -0,0 +1,789 @@
+automate it.** This allows us to onboard people into the project by doing a
+simple `git clone` and `bin/setup`. **Time to first contribution is a couple of
+hours instead of 1+ days**.
+
+Let's get right into the basic scripts, the advanced scripts and the insane ones.
+
+## Into the basics
+
+As I mentioned, there are three essential scripts: `bin/server`, `bin/setup` and
+`bin/lint`. The first runs the development server. The second installs
+everything you might need to run the project. The third runs all linters and
+checks, usually being equivalent to running the CI locally.
+
+I use these so much, both at Subvisual and on personal projects, I have aliases
+for all of them and I recommend you do the same.
+
+### bin/server
+
+Usually the simplest script, these are examples for Elixir, Ruby and JavaScript
+(Gatsby) projects:
+
+```shell
+#!/usr/bin/env sh
+
+# Elixir
+iex -S mix phx.server
+
+# Ruby
+bundle exec foreman start -f Procfile.dev
+
+# JavaScript / Gatsby
+yarn start
+```
+
+Sometimes, however, we need to do complex operations. As an example, in an
+internal project we had to set up `ngrok` before starting the server. Because
+this is something we tend to forget and newcommers might not be aware of, I made
+this `bin/` script that launches `ngrok` in the background, gets generated URL,
+launches the server and kills `ngrok` before exiting.
+
+```shell
+#!/usr/bin/env bash
+
+trap ctrl_c INT
+
+ctrl_c() {
+  printf "\n"
+  echo "Received kill signal, stopping ngrok... "
+  $(killall ngrok &>/dev/null);
+
+  printf "Done!\n"
+
+  exit 1
+}
+
+PORT=${PORT:-4000}
+
+echo "Starting ngrok... "
+
+ngrok http $PORT > /dev/null &
+
+sleep 2
+
+REGEX="\"public_url\":\s*\"(https://([^\"]+))\"[^{]+{\"addr\":\s*\"http:\/\/localhost:$PORT\""
+
+JSON_RESPONSE=$(curl http://localhost:4040/api/tunnels 2>/dev/null)
+
+if [[ "$JSON_RESPONSE" =~ $REGEX ]]; then
+  export WEBHOOK_URL=${BASH_REMATCH[1]}
+  export HOST=${BASH_REMATCH[2]}
+  export ENDPOINT_PORT="443"
+  export SCHEME="https"
+
+  printf "Done!\n\n"
+  echo "Your webhook url is $WEBHOOK_URL"
+else
+  echo "Wow, something went wrong with ngrok!";
+  ctrl_c
+fi
+
+echo "Starting development server with options HOST=$HOST SCHEME=$SCHEME ENDPOINT_PORT=$ENDPOINT_PORT..."
+
+iex -S mix
+```
+
+<a href="https://asciinema.org/a/Jbbe7Fvt5OYHEvqAfqEAyJeXc" target="_blank">
+  <img src="https://asciinema.org/a/Jbbe7Fvt5OYHEvqAfqEAyJeXc.svg" alt="bin/server in action" style="max-width: 100%"/>
+</a>
+
+Anyone coming into the project doesn't need to be aware of this complexity, they
+would just run a command and voilá. Magic.
+
+### bin/lint
+
+Usually nothing more than the scripts you would run on CI. As an example, here's
+what we have on a project that runs the linters for frontend (in JavaScript) and
+backend (in Elixir). Notice the `set -e`. This causes the script to exit if a
+command has a non-zero exit code. The reasoning behind this is to mimic the CI
+behaviour.
```

Comment:
so the `-e` is here after all? You know it's suposed to go on every other file as well right?

Also, `set -e` is equivalent to `#!/usr/bin/env sh -e`, ur just `#!/bin/sh -e` directly